### PR TITLE
ARROW-6637: [C++] Further streamline default build, add ARROW_CSV CMake option

### PR DESCRIPTION
**Pull request by wesm**
* Default for ARROW_COMPUTE/DATASET/FILESYSTEM/JSON set to OFF
* Add ARROW_CSV option, set to OFF by default (I could be swayed about this, but I think it is good to not create the perception that users are forced to build a module they don't need)
* Disable unit tests that don't build if ARROW_COMPUTE=OFF

The minimal Docker build now outputs

https://gist.github.com/wesm/ff70f46f5bc256d6b1d7b979aab8e5d0

The minimal build time on my 8-core laptop is reduced from 72 seconds to 25 seconds. Here is the log for the default build (with jemalloc turned off) without this patch

https://gist.github.com/wesm/fb5300a1989420158b0148e1c49ab6d6

Note that the default for ARROW_JEMALLOC is left to ON. I agree based on other discussion that having it off by default might cause developers doing performance testing to draw the wrong conclusions due to worse performance from the system allocator. 

TODO

- [ ] Check conda
- [ ] Check Homebrew
- [ ] Check wheels
- [ ] Check Linux packages

### CODE REVIEW

**Comment by wesm** [comment]
@ursabot crossbow submit -g wheel

**Comment by wesm** [comment]
@kszucs The Ursabot builds are going to fail here I think

**Comment by wesm** [comment]
@kszucs it's weird that the manylinux1 builds are building LLVM from source, is something wrong with the cached Docker image?

**Comment by wesm** [comment]
@ursabot crossbow submit -g wheel -g linux -g conda -g homebrew

**Comment by wesm** [comment]
@ursabot crossbow submit -g wheel -g linux -g conda

**Comment by wesm** [comment]
This is ready for review / merge once the package builds are all working. I'll work on a follow up patch to improve optional module / dependency management, something like `-DARROW_BUILD_MODULES=csv,dataset,json,parquet`

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -141,20 +141,22 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
...
-  define_option(ARROW_FILESYSTEM "Build the Arrow Filesystem Layer" ON)
+  define_option(ARROW_FILESYSTEM "Build the Arrow Filesystem Layer" OFF)
 
   define_option(ARROW_FLIGHT
                 "Build the Arrow Flight RPC System (requires GRPC, Protocol Buffers)" OFF)
 
   define_option(ARROW_GANDIVA "Build the Gandiva libraries" OFF)
 
-  define_option(ARROW_HDFS "Build the Arrow HDFS bridge" ON)
+  define_option(ARROW_HDFS "Build the Arrow HDFS bridge" OFF)
```
Comment (kou) [inline cpp/cmake_modules/DefineOptions.cmake]: It seems that adding `-DARROW_HDFS=ON` explicitly is missing.

Comment (wesm) [reply cpp/cmake_modules/DefineOptions.cmake]: I see, thanks, will fix

**Comment by nealrichardson** [comment]
I'm aware of a few more places you'll want to turn these flags back `ON` or otherwise update docs: 

* https://github.com/apache/arrow/blob/master/dev/tasks/homebrew-formulae/apache-arrow.rb#L26
* https://github.com/apache/arrow/blob/master/dev/tasks/homebrew-formulae/autobrew/apache-arrow.rb#L42
* https://github.com/apache/arrow/blob/master/ci/PKGBUILD#L79
* https://github.com/apache/arrow/blob/master/docs/source/developers/cpp.rst

**Comment by wesm** [comment]
@nealrichardson I'll take a look, but not that `ARROW_PYTHON=ON` implies COMPUTE, CSV, DATASET, FILESYSTEM, HDFS, JSON are all on, so it's not necessary to put these everywhere

**Comment by nealrichardson** [comment]
`ARROW_PYTHON` is on for homebrew-cpp but not for the other places I noted, and in any case that sounds like something that deserves documentation.

**Comment by wesm** [comment]
Yes, will do.

**Comment by wesm** [comment]
@ursabot crossbow submit -g wheel -g linux -g conda

**Comment by wesm** [comment]
@pitrou @kszucs this is failing for the same reason that the manylinux nightlies are failing, due to the cached Docker image not properly downloading. 

I checked the manylinux builds locally and all looking good there. 

This is ready to merge

**Review by nealrichardson — CHANGES_REQUESTED** [review]
One note; also need to add those flags to https://github.com/apache/arrow/blob/master/dev/tasks/homebrew-formulae/autobrew/apache-arrow.rb.

**Comment by nealrichardson** [comment]
@ursabot crossbow submit macos-r-autobrew

**Comment by kszucs** [comment]
I'll wait until tomorrow for the apache dockerhub access, if we get access then I'm going to push the manylinux images, otherwise I'm going to fall back to the previous docker org for the manylinux images.

File: ci/PKGBUILD [hunk 2]
```diff
@@ -80,6 +80,12 @@ build() {
     -DCMAKE_BUILD_TYPE=${cmake_build_type} \
     -DARROW_BUILD_STATIC=ON \
     -DARROW_BUILD_SHARED=OFF \
+    -DARROW_COMPUTE=ON \
+    -DARROW_CSV=ON \
+    -DARROW_DATASET=ON \
+    -DARROW_FILESYSTEM=ON \
+    -DARROW_HDFS=ON \
```
[the remark this answers is not in the document]
Comment (wesm) [reply ci/PKGBUILD]: fixed

**Comment by wesm** [comment]
I don't think the DockerHub issue should block this patch

**Comment by kou** [comment]
I've merged https://github.com/apache/arrow/pull/5966 and rebased this on master.

**Comment by kou** [comment]
@ursabot crossbow submit -g wheel

**Review by kou — APPROVED** [review]
+1
We need to update ursabot configurations for the followings:

* https://ci.ursalabs.org/#/builders/97/builds/1724
* https://ci.ursalabs.org/#/builders/96/builds/3044

But it's out of scope of this. Because ursabot configuration isn't in this repository.

(Can we stop ursabot based CI?)

**Comment by kszucs** [comment]
@kou I've disabled the responsible schedulers

**Comment by kou** [comment]
Thanks!

**Pull request closed without merging**